parser/src/parser/file.rs

```diff
@@ -359,6 +373,31 @@ impl ParserContext {
         Ok(members)
     }
 
+    ///
+    /// Returns a [`CircuitMember`] AST node if the next tokens represent a circuit member static constant.
+    ///
+    pub fn parse_const_member_variable_declaration(&mut self) -> Result<(CircuitMember, bool, bool)> {
+        self.expect(Token::Static)?;
+        self.expect(Token::Const)?;
```

Comment:
I'm assuming `static const` was the spec. However, the extra `static` seems odd to me. In Rust, C, and friends, a `static` means the variable gets assigned a unique and global memory address. It seems to me that's not the case here, so why not just `const IDENT: TYPE = EXPR;`?

Reply:
Part of the reason is that the access itself is done through static access rather than nonstatic access. So unless we also changed the accessor from `::` to `.`. Some other context can be found in https://github.com/AleoHQ/leo/blob/master/docs/rfc/011-scalar-type-accesses-and-methods.md.

Reply:
We also wanted to leave the door open to non-static member constants, as a possible future extensions. These would be per-instance, initialized with `const` expressions, and never changed of course, so they can be partially evaluated away -- every variable and constraint we can save in R1CS counts.

Reply:
@gluax Sure but we don't say:
```rust
circuit Foo {
    static function bar() {}
}

Foo::bar();
```
so that reasoning seems inconsistently applied.

@acoglio So that means allowing this?

```rust
circuit Foo { const x: u8; }

function bar(const y: u8) {
    let f = Foo { x: y };
    f.x;
}
```

If so, this rather suggests that `const` should be a modifier on *types* i.e., `const u8` instead, with subtyping:
```
Γ |- σ type
----------------- ConstSubtypesType
Γ |- const σ <: σ
```

You'll also need syntactic conveniences like `const [u8; N]` being accepted meaning `const [const u8; N]`.
Ostensibly, this system should be more flexible and more orthogonal, but please try to poke holes in it.

Reply:
I think it's an interesting discussion. It feels slightly strange, mostly because I haven't seen the `const` modifier applied to types directly like that before. I agree it seems more orthogonal at first glance, so I suppose the question is the less familiar syntax for the user worth it?

Reply:
Yeah, maybe let's make an issue out of this and discuss at a later time?

Reply:
@Centril Yes, there is an inconsistency (which bothers me slightly as well) in regard to not using `static` for functions. Early versions of Leo actually had that, but it was changed in favor of Rust's "either has a `self` parameter or not" approach. With static constants, we have a mix of Rust's and Java's styles.

Something else may actually resolve this `static` inconsistency between functions and constants. The issue https://github.com/AleoHQ/leo/issues/1417 advocates putting circuit member functions outside circuit type declarations, based on the idea that member functions are not only for circuit types, but for other kinds of types as well. The same applies to static constants and possibly static variables, as the last line of that Issue says. I'm quite sure we'll eventually implement https://github.com/AleoHQ/leo/issues/1417. Then the things inside the `members` or `impl` block would not need any `static` modifier, because they are all implicitly associated to the type (not to specific instances/values of the type).

I think that making `const` a type modifier, along with the subtyping expressed by your rule, is an idea worth considering, and it would be great if you could open an issue about that as you suggest. (@gluax This is somewhat analogous to C's `const`, which is classified as a 'type qualifier'.)

Reply:
Greated https://github.com/AleoHQ/leo/issues/1447 to track the discussion.